Feat/comments

ehmadzubair opened the pull request:
This PR fixes #56 

Here's a list of the features done as part of this:

> - Commenting can be enabled or disabled on any page. 

This is done for the Page and its comments as well. Disallowing comments hides both the "New Comment" form and the "Reply" links on existing comments

> - Users can post comments

Go to Article screen and scroll to the bottom part of the page. Make sure that "Allow Comments" is set to true on Admin > Pages > Article > Settings

> - Users can 'like' comments

Done

> - Moderators can review, filter and export comments. 

See Admin > Comments > All Comments. Look for button on hovering the comments.

> - Moderators can provide custom or canned responses to comments.

See Admin > All Comments > Reply button on a comment

> - Community moderators can flag inappropriate comments.

Everybody can flag comments for now but only moderators can delete comments. Since flagging is also a requirement for frontend users, the same feature is shared between them. Thoughts @dannyparsons @istride?

> - Moderators should be alerted to flagged comments.

Discussed with @dannyparsons that this can be replaced with a filter. The filter can be found here "Admin > Comments > All Comments > Flagged Filter"

> - Fix #33 

All the requirements for #33 are complete except for comment sorting. This PR was becoming very lengthy, so I decided to merge this first and then do the minor changes in another PR

> - Fix #34 

Flagging is the same as reporting. 

> - Fix #35 

Done. Add canned responses from Admin > Comments > Canned Responses

> - #37 

This will have to be done in a separate PR


## Testing Instructions
There are no special steps required for testing. Just running the migrations and running the app should show up the comments box and comments list on the frontend. The admin panel should have a new sections named "Comments" on the sidebar


Note: Created #86 for issues that are not completed in this

istride commented on comments/tests.py:
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
+
+# Create your tests here.
```
would rather omit this file, if there are no tests

istride commented on home/templates/home/article.html:
```diff
@@ -1,8 +1,11 @@
-<link rel="stylesheet" type="text/css" href="{% static 'css/article.css' %}">
+    <link rel="stylesheet" type="text/css" href="{% static 'css/article.css' %}">
+    <link rel="stylesheet" type="text/css" href="{% static 'css/comments.css' %}">
+    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css" integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU" crossorigin="anonymous">
```
I think we'll have to insist that all assets are served by IoGT to ensure that users don't incur data charges. The addition of font awesome may be temporary, but I would prefer that we always make sure that we host everything ourselves at all times.

istride commented on home/models.py:
```diff
@@ -90,11 +91,18 @@ def get_paragraph_values(self):
         paragraph_values = self._get_child_block_values('paragraph')
         return '\n'.join(paragraph_values)
 
+    def get_absolute_url(self):
```
not sure why this is necessary

ehmadzubair replied:
The `comments_xtd` library uses the `et_absolute_url` function of the model to redirect to after every successful action.

ehmadzubair replied:
Move this to the Mixin

istride commented on comments/models.py:
```diff
@@ -0,0 +1,19 @@
+from django_comments_xtd.models import XtdComment
+
+
+class AllowCommentsModelMixin(models.Model):
```
would consider renaming this to `CommentsMixin` or `CommentableMixin`.

what happens if other attributes are required in this class e.g. time-restricted commenting?

ehmadzubair replied:
Shouldn't 'other required attributes' be set in the setting panels?

istride commented on comments/views.py:
```diff
@@ -0,0 +1,69 @@
+from comments.models import CannedResponse
+
+
+def update(request, comment_pk, action):
```
My code smell senses are tingling, but I am not able to offer a viable alternative. Thinking from a RESTful perspective, I would want each 'action' here to be handled by a separate endpoint and this would eliminate the 'if-elif' structure here.

If no change is made, should there be some tests to catch possible regressions in the future if someone has to edit this code?

ehmadzubair replied:
This code has been 'copied' from `wagtail_comments_xtd`. Will discuss in the call

istride commented on home/models.py:
```diff
@@ -90,11 +91,18 @@ def get_paragraph_values(self):
     ]
 
+    settings_panels = Page.settings_panels + [
+        FieldPanel('allow_comments')
```
My preference would be for the panel to be titled 'Comments', and for there to be a checkbox labelled 'Enabled' within the panel.

istride reviewed (changes requested):
A few more things that I noticed:

- every interaction with the comments system, as a user, redirects me to 'example.com'
- not a fan of asking for confirmation when liking comments
- i can still see the old comments template on article pages (is this a case of clearing browser cache?)
- if comments are switched off, I think there should be no mention of comments anywhere
- as an admin editing a comment, there a lot of attributes that can be changed, which perhaps, should not be allowed to be changed, such as 'content_type', 'object_id' - admins could get into real trouble by fiddling with these.

ehmadzubair commented on comments/admin.py:
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
+
+# Register your models here.
```
This is auto-generated by Django. Will do

ehmadzubair commented on iogt/settings/profanity_settings.py:
```diff
@@ -0,0 +1,2 @@
+COMMENTS_ALLOW_PROFANITIES = False
+PROFANITIES_LIST = ('crap', 'hell')
```
Discuss with @istride

ehmadzubair commented:
@istride 

> every interaction with the comments system, as a user, redirects me to 'example.com'

I couldn't reproduce this. 

> not a fan of asking for confirmation when liking comments

Will figure out how to get rid of this (default behaviour of comments-xtd)

> i can still see the old comments template on article pages (is this a case of clearing browser cache?)

This was pushed later in `02ee0de`. You shoulndn't see it now.

> if comments are switched off, I think there should be no mention of comments anywhere

@dannyparsons and I had discussed that setting allow comments to false would only stop new responses. Do we want to hide all the comments with the same flag OR create a new flag for this?

> as an admin editing a comment, there a lot of attributes that can be changed, which perhaps, should not be allowed to be changed, such as 'content_type', 'object_id' - admins could get into real trouble by fiddling with these.

Good catch. Will limit the fields

dannyparsons commented:
On disabling comments, I think we should keep existing comments visible if they have been made. That's what done in v1 I think and I haven't seen a request to change this. I notice that on articles that have never had comments there's no mention of comments https://sa.goodinternet.org/sections/climate-change/solutions/renewable-energy/ so good to keep that functionality if possible.

ehmadzubair commented:
> On disabling comments, I think we should keep existing comments visible if they have been made. That's what done in v1 I think and I haven't seen a request to change this. Comments can be deleted if needed separately.
> 
> I notice that on articles that have never had comments there's no mention of comments https://sa.goodinternet.org/sections/climate-change/solutions/renewable-energy/ so good to keep that functionality if possible.

Got it. I've changed the functionality to be exactly the same

ehmadzubair commented:
@istride All the changes that were requested have been done. I've also incorporated all the previous feedback except for the test cases. I've added that to #86

istride commented:
@ehmadzubair would you limit font awesome to the bare minimum necessary to make your feature work, or just leave it out altogether. I'm not sure what the new design will be like and whether it will require font awesome or not. I certainly don't want to commit nearly 2000 files, only to take them away a few days later, just for the sake of a few icons.

ehmadzubair commented:
> @ehmadzubair would you limit font awesome to the bare minimum necessary to make your feature work, or just leave it out altogether. I'm not sure what the new design will be like and whether it will require font awesome or not. I certainly don't want to commit nearly 2000 files, only to take them away a few days later, just for the sake of a few icons.

Yes, will do. I'll do the latter.

ehmadzubair commented:
@isedwards, @istride and I went through this PR in detail and there were two changes that were pending yesterday. I have made both of them. Can you merge this soon? (it has already caused a lot of diverging migration trees)

cc: @dannyparsons
